### PR TITLE
ChannelForm: Render default values as element values

### PR DESCRIPTION
**Pull request by raviks789**
Instead of rendering default values as placeholders, render them as element values.

### CODE REVIEW

File: application/forms/ChannelForm.php [hunk 1]
```diff
@@ -188,17 +198,25 @@ protected function createConfigElements(string $type, string $config): void
...
                 $elem->getValidators()->add(new EmailAddressValidator());
             }
 
+            if (
+                isset($elementConfig->default)
+                && $configFieldset->getPopulatedValue($elementConfig->name) === null
+            ) {
+                $configFieldset->clearPopulatedValue($elementConfig->name);
+            }
+
```
Comment (nilmerg) [inline application/forms/ChannelForm.php:219]: Please remove and avoid that null values are stored in the database instead.

File: application/forms/ChannelForm.php [hunk 2]
```diff
@@ -188,17 +198,25 @@ protected function createConfigElements(string $type, string $config): void
         $this->addElement($configFieldset);
 
         foreach ($elementsConfig as $elementConfig) {
+            $elemOptions = $this->getElementOptions($elementConfig);
             /** @var BaseFormElement $elem */
             $elem = $this->createElement(
                 $this->getElementType($elementConfig),
                 $elementConfig->name,
-                $this->getElementOptions($elementConfig)
+                $elemOptions
```
Comment (nilmerg) [inline application/forms/ChannelForm.php:206]: Please undo these redundant changes

File: application/forms/ChannelForm.php [hunk 3]
```diff
@@ -161,7 +163,20 @@ protected function onSuccess()
...
+            function ($configItem) {
+                return $configItem !== null;
+            }
+        );
+
+        foreach ($this->defaultChannelOptions as $key => $value) {
+            if (array_key_exists($key, $config) && $config[$key] === $value) {
+                unset($config[$key]);
+            }
+        }
```
Comment (nilmerg) [inline application/forms/ChannelForm.php:177]: You can do both things with array_filter

**Review by nilmerg — CHANGES_REQUESTED** [review]

File: application/forms/ChannelForm.php [hunk 4]
```diff
@@ -177,8 +206,10 @@ protected function onSuccess()
      */
     protected function createConfigElements(string $type, string $config): void
     {
-        /** @var array<int, stdClass> $elementsConfig */
-        $elementsConfig = json_decode($config, false);
+        /**
+         * @var array<int, ChannelOptionConfig>  $elementsConfig
+         */
```
Comment (nilmerg) [inline application/forms/ChannelForm.php:211]: single line still, pls

File: application/forms/ChannelForm.php [hunk 5]
```diff
@@ -204,15 +235,15 @@ protected function createConfigElements(string $type, string $config): void
     }
 
     /**
-     * Get the element type from given element config
+     * Get the element type config type of the object
```
Comment (nilmerg) [inline application/forms/ChannelForm.php:238]: *Get the element type for the given option type*

File: application/forms/ChannelForm.php [hunk 6]
```diff
@@ -204,15 +235,15 @@ protected function createConfigElements(string $type, string $config): void
...
      *
-     * @param stdClass $elementConfig The config object of an element
+     * @param string $configType The config type of the object
```
Comment (nilmerg) [inline application/forms/ChannelForm.php:240]: *The option type*

**Review by nilmerg — CHANGES_REQUESTED** [review]

File: application/forms/ChannelForm.php [hunk 7]
```diff
@@ -204,15 +233,15 @@ protected function createConfigElements(string $type, string $config): void
     }
 
     /**
-     * Get the element type from given element config
+     * Get the element type for the given option type
      *
-     * @param stdClass $elementConfig The config object of an element
+     * @param string $configType The option type of the object
```
[the remark this answers is not in the document]
Comment (raviks789) [reply application/forms/ChannelForm.php]: Supposed to be _The option type_. I have corrected it.

**Review by nilmerg — COMMENTED** [review]
Looks fine now. But I don't like the handling of defaults for required elements: https://github.com/Icinga/icinga-notifications/issues/172

**Review by nilmerg — CHANGES_REQUESTED** [review]
Default values, that are not adjusted by the user, **must not** be written to database.

**Review by nilmerg — APPROVED** [review]

**Pull request merged**